### DISCUSSION TITLE
Legacy tables in Emmett ORM

### DISCUSSION THREAD

**Discussion by denesl** [emmett-framework/emmett#391] [Q&A] [answered]
Hi

could you please post an example on how to define legacy tables in Emmett?

In pydal you would use the primarykey parameter in define_table, and maybe the rname (real table name in the db).
primarykey might be empty, [], if there is no PK.

orm.migrations.base.py seems to limit PK to only one field, correct?.

Thanks.

**Comment by gi0baro** [comment, marked as answer]
It's not yet supported, but it will in the next release (2.4).
It's actually implemented with https://github.com/emmett-framework/emmett/pull/385, but I still have to write docs.

You can try the [2.4 branch](https://github.com/emmett-framework/emmett/tree/2.4), using the `primary_keys` attribute over your models, eg:

```python
from emmett.orm import Model, Field

class MyModel(Model):
    primary_keys = ["field_a", "field_b"]

    field_a = Field()
    field_b = Field()
```

**Reply by denesl** [reply]
I would like to test with 2.4 but I am not sure how to start emmett .
I got as far as having Emmett-2.4.0.dev0-py3.9.egg in a python 3.9.7 venv .
pip list shows it as Emmett 2.4.0.dev0 .
Thank you.

**Reply by gi0baro** [reply]
If you installed 2.4 in a venv, just activate it from the shell and you can use the `emmett` command as usual.

**Reply by denesl** [reply]
If I try calling emmett I get:

> (emmett24) C:\Python39\venv\emmett24>emmett
> 'emmett' is not recognized as an internal or external command,
> operable program or batch file.

This is what pip shows:
```
(emmett24) C:\Python39\venv\emmett24>python -m pip list
Package    Version
---------- ----------
Emmett     2.4.0.dev0
pip        21.3.1
setuptools 58.3.0
```

**Reply by gi0baro** [reply]
@denesl how did you installed it? I would run `pip install git+https://github.com/emmett-framework/emmett.git@2.4`, eg:

```
giovanni@Deneb ~/Desktop ❯ mkdir -p emt24test
giovanni@Deneb ~/Desktop ❯ cd emt24test
giovanni@Deneb ~/Desktop/emt24test ❯ python -m venv .venv
giovanni@Deneb ~/Desktop/emt24test ❯ source .venv/bin/activate
(.venv) giovanni@Deneb ~/Desktop/emt24test ❯ pip install git+https://github.com/emmett-framework/emmett.git@2.4
Collecting git+https://github.com/emmett-framework/emmett.git@2.4
  Cloning https://github.com/emmett-framework/emmett.git (to revision 2.4) to /private/var/folders/d3/bsymrzt9651dx7gp9yxb4qmw0000gn/T/pip-req-build-1wqp9dzi
  Running command git clone -q https://github.com/emmett-framework/emmett.git /private/var/folders/d3/bsymrzt9651dx7gp9yxb4qmw0000gn/T/pip-req-build-1wqp9dzi
  Running command git checkout -b 2.4 --track origin/2.4
  Switched to a new branch '2.4'
  Branch '2.4' set up to track remote branch '2.4' from 'origin'.
  Resolved https://github.com/emmett-framework/emmett.git to commit ae9f2f9d6212a7a49d65bb406ea82f1c3bd788d3
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
    Preparing wheel metadata ... done
Collecting uvloop<0.17.0,>=0.16.0
  Using cached uvloop-0.16.0-cp310-cp310-macosx_10_9_x86_64.whl (1.5 MB)
Collecting pyyaml<6.0,>=5.4
  Using cached PyYAML-5.4.1-cp310-cp310-macosx_10_14_x86_64.whl
Collecting pendulum<2.2.0,>=2.1.2
  Using cached pendulum-2.1.2-cp310-cp310-macosx_10_14_x86_64.whl
Collecting pyaes<1.7.0,>=1.6.1
  Using cached pyaes-1.6.1.tar.gz (28 kB)
Collecting python-rapidjson<2.0,>=1.0
  Using cached python_rapidjson-1.5-cp310-cp310-macosx_10_9_x86_64.whl (214 kB)
Collecting uvicorn==0.15.0
  Using cached uvicorn-0.15.0-py3-none-any.whl (54 kB)
Collecting httptools<0.4.0,>=0.3.0
  Using cached httptools-0.3.0-cp310-cp310-macosx_10_9_x86_64.whl (156 kB)
Collecting pyDAL==17.3
  Using cached pyDAL-17.3-py2.py3-none-any.whl (169 kB)
Collecting severus<2.0,>=1.1
  Using cached Severus-1.1.1-py3-none-any.whl (12 kB)
Collecting websockets<10.0,>=9.1
  Using cached websockets-9.1.tar.gz (76 kB)
Collecting renoir<2.0,>=1.5
  Using cached Renoir-1.5.0-py3-none-any.whl (29 kB)
Collecting h11<0.13.0,>=0.12.0
  Using cached h11-0.12.0-py3-none-any.whl (54 kB)
Collecting h2<4.1.0,>=3.2.0
  Using cached h2-4.0.0-py3-none-any.whl (57 kB)
Collecting click>=6.0
  Using cached click-8.0.3-py3-none-any.whl (97 kB)
Collecting asgiref>=3.4.0
  Using cached asgiref-3.4.1-py3-none-any.whl (25 kB)
Collecting hyperframe<7,>=6.0
  Using cached hyperframe-6.0.1-py3-none-any.whl (12 kB)
Collecting hpack<5,>=4.0
  Using cached hpack-4.0.0-py3-none-any.whl (32 kB)
Collecting python-dateutil<3.0,>=2.6
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting pytzdata>=2020.1
  Using cached pytzdata-2020.1-py2.py3-none-any.whl (489 kB)
Collecting six>=1.5
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Using legacy 'setup.py install' for pyaes, since package 'wheel' is not installed.
Using legacy 'setup.py install' for websockets, since package 'wheel' is not installed.
Building wheels for collected packages: emmett
  Building wheel for emmett (PEP 517) ... done
  Created wheel for emmett: filename=Emmett-2.4.0.dev0-py3-none-any.whl size=443467 sha256=040669fb6252d0ca37c5fa7e4ed9cd96f916003c91c62806528afd3ffd4304bf
  Stored in directory: /private/var/folders/d3/bsymrzt9651dx7gp9yxb4qmw0000gn/T/pip-ephem-wheel-cache-er55r1sf/wheels/6f/2a/3c/33e5d2064e580998fb0adbc3289a9ae61337843a4c271e2b00
Successfully built emmett
Installing collected packages: six, pyyaml, pytzdata, python-dateutil, hyperframe, hpack, h11, click, asgiref, websockets, uvloop, uvicorn, severus, renoir, python-rapidjson, pyDAL, pyaes, pendulum, httptools, h2, emmett
    Running setup.py install for websockets ... done
    Running setup.py install for pyaes ... done
Successfully installed asgiref-3.4.1 click-8.0.3 emmett-2.4.0.dev0 h11-0.12.0 h2-4.0.0 hpack-4.0.0 httptools-0.3.0 hyperframe-6.0.1 pendulum-2.1.2 pyDAL-17.3 pyaes-1.6.1 python-dateutil-2.8.2 python-rapidjson-1.5 pytzdata-2020.1 pyyaml-5.4.1 renoir-1.5.0 severus-1.1.1 six-1.16.0 uvicorn-0.15.0 uvloop-0.16.0 websockets-9.1
WARNING: You are using pip version 21.2.3; however, version 21.3.1 is available.
You should consider upgrading via the '/Users/giovanni/Desktop/emt24test/.venv/bin/python -m pip install --upgrade pip' command.
(.venv) giovanni@Deneb ~/Desktop/emt24test ❯ which emmett
/Users/giovanni/Desktop/emt24test/.venv/bin/emmett
(.venv) giovanni@Deneb ~/Desktop/emt24test ❯ emmett --help
Usage: emmett [OPTIONS] COMMAND [ARGS]...

Options:
  -a, --app TEXT  The application to run
  --help          Show this message and exit.

Commands:
  develop     Runs a development server.
  migrations  Runs migration operations.
  routes      Display the app routing table.
  serve       Serve the app.
  shell       Runs a shell in the app context.
(.venv) giovanni@Deneb ~/Desktop/emt24test ❯
```

**Reply by denesl** [reply]
Recreated the venv and ran pip install c:/temp/emmet-2.4 (where emmett-2.4.zip is unpacked). 
The call to start emmett works now. Thank you.

Back to the original question: how would I supply the legacy table name?
In pydal I would use rname= '[INFORMATION_SCHEMA].[TABLES]'

**Reply by gi0baro** [reply]
There's no *direct way* to do that in Emmett, but you can manually change the relevant attribute in the `setup` method of your model, eg:

```python
class MyModel(Model):
    def setup(self):
        self.table._rname = '"myschema"."mytable"'
```

**Reply by denesl** [reply]
Clever. I tried it but I am getting an error. It seems that it thinks there should be an id field. 

```
from emmett.orm import Database, Model, Field

app = App(__name__)

app.config.db.uri = r'mssql://user:pass@server/dbname'

class tablesinfo(Model):
  primary_keys = []
  TABLE_NAME = Field(length=128,notnull=True)
  TABLE_TYPE = Field(length=10)
  def setup(self):
    self.table._rname = '[INFORMATION_SCHEMA].[TABLES]'

db = Database(app)

db.define_models(tablesinfo)

rr=db(tablesinfo.TABLE_NAME=='XYZ').select()
```

> Traceback (most recent call last):
>   File "<console>", line 1, in <module>
>   File "C:\Python39\venv\emmett24\lib\site-packages\emmett\orm\objects.py", line 345, in select 
>     return obj._run_select_(*fields, **options)
>   File "C:\Python39\venv\emmett24\lib\site-packages\emmett\orm\objects.py", line 329, in _run_select_
>     return super(Set, self).select(*fields, **options)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\objects.py", line 2210, in select
>     return adapter.select(self.query, fields, attributes)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\adapters\base.py", line 760, in select
>     return self._select_aux(sql, fields, attributes, colnames)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\adapters\base.py", line 716, in _select_aux
>     rows = self._select_aux_execute(sql)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\adapters\base.py", line 710, in _select_aux_execute
>     self.execute(sql)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\adapters\__init__.py", line 67, in wrap
>     return f(*args, **kwargs)
>   File "C:\Python39\venv\emmett24\lib\site-packages\pydal\adapters\base.py", line 410, in execute
>     rv = self.cursor.execute(command, *args[1:], **kwargs)
> pyodbc.ProgrammingError: ('42S22', "[42S22] [Microsoft][ODBC SQL Server Driver][SQL Server]Invalid column name 'id'. (207) (SQLExecDirectW)")

I can get around that by listing the fields in the select:
`rr=db(tablesinfo.TABLE_NAME=='XYZ').select(tablesinfo.TABLE_NAME, tablesinfo.TABLE_TYPE)`

**Reply by gi0baro** [reply]
@denesl that's not how it is supposed to work.

The ORM needs a Model identifier, so there's no way to suppress primary keys. If you set `primary_keys = []` then the ORM will automatically define an `id` field.

**Reply by denesl** [reply]
@gi0baro  I have used primarykey=[] since the support for legacy tables was added to web2py in 2009, before pyDAL.
Otherwise it would not be possible to access legacy tables that do not have a primary key.
How do I know? I wrote the original code ;)

It still works today with the latest pydal e.g.
```
>>> from pydal import DAL,Field
>>> db=DAL(r'mssql://user:pass@server/dbname', migrate=False)
>>> db.define_table('tablesinfo',
...   Field('TABLE_NAME', 'string', 128, notnull=True),
...   Field('TABLE_TYPE', 'string', 10),
...   rname='[INFORMATION_SCHEMA].[TABLES]',
...   primarykey=[],
...   )
<Table tablesinfo (TABLE_NAME, TABLE_TYPE)>
>>> tt = db(db.tablesinfo.TABLE_NAME == 'WXYZ').select()
>>> print(tt)
tablesinfo.TABLE_NAME,tablesinfo.TABLE_TYPE
WXYZ,BASE TABLE
```

**Reply by gi0baro** [reply]
@denesl yes, of course it works with pyDAL: pyDAL is not an ORM indeed.
ORMs need identifiers for records, otherwise 90% of their features won't just be available.

To get back to your original question: if you want to use tables without an "id" column, or multiple primary keys, then with Emmett 2.4 you can. If your need is to use tables without primary keys in Emmett's ORM (and I'm pretty sure in any other ORM out there), then it won't never happen, because the ORM needs such information _by design_.

In the latest snippet you shared, I think you can trick it with something like `primary_keys = ["TABLE_NAME", "TABLE_TYPE"]`, but I can't foresee eventual further implications.

**Comment by denesl** [comment]
Hi gi0baro, I understand what you say about ORMs.
But the point is that pyDAL provides a way to use legacy tables, and there is still a lot of them out there, with or without key(s).
Note that they can only interact with other legacy tables, so no mixing with tables with an id field.
And that is a big plus, at least in my book.

I am a bit confused by the last 2 paragraphs of your previous answer, In emmett 2.4 it is OK to have:
- tables without an id column (when there is another field that can play the id role or when an id field can be added to the table, right?)
- tables with multiple primary keys (as long as one acts as the id, otherwise things might go sour at some point, right?)

Sorry to keep harping on but I need to know if I am using the proper tool for the job.

**Reply by gi0baro** [reply]
@denesl no worries, is always nice to receive questions and feedbacks :)

Btw, to clarify:

> tables without an id column (when there is another field that can play the id role or when an id field can be added to the table, right?)

Correct, you need to have at least a primary key.

> tables with multiple primary keys (as long as one acts as the id, otherwise things might go sour at some point, right?)

Nope, with multiple primary keys the ORM *should* be able to operate using the two of them as the `id` field. Of course this need more testing, but this is the idea.

All this considered, I'm not sure the ORM is the proper tool for you needs, but you can still use the pyDAL layer from the ORM in Emmett, as the `Database` class in Emmett's ORM is actually a subclass of `DAL`, so you still have the `define_table` method, and you can still use the `db.pipe` object to handle connections in your application flow. I think this would work better for your use case.
